File: ts/packages/trellis/telemetry/lifecycle.ts
import {
  recordTrellisDuration,
  type TrellisDurationMetricAttributes,
  type TrellisDurationMetricName,
} from "./metrics.ts";

/** Exactly-once duration observation for one unit of work. */
export interface TelemetryObservation {
  /** Records the observation with one bounded outcome value. */
  finish(outcome: string, extra?: TrellisDurationMetricAttributes): void;
  /** Records the configured cancelled/interrupted outcome once. */
  cancel(): void;
}

/**
 * Starts one duration observation.
 *
 * Callers use `try/finally` so `?`/throw paths cannot silently skip the
 * observation, matching the Rust observation guard boundary.
 */
export function startObservation(
  metric: TrellisDurationMetricName,
  attributes: TrellisDurationMetricAttributes,
  dropOutcome = "cancelled",
): TelemetryObservation {
  const startedAt = performance.now();
  let finished = false;
  const record = (
    outcome: string,
    extra?: TrellisDurationMetricAttributes,
  ): void => {
    if (finished) return;
    finished = true;
    recordTrellisDuration(
      metric,
      performance.now() - startedAt,
      { ...attributes, ...extra, outcome },
    );
  };
  return {
    finish: (outcome, extra) => record(outcome, extra),
    cancel: () => record(dropOutcome),
  };
}

/**
 * Runs one async unit under a duration observation with the same boundary.
 *
 * The original value or exception is always preserved.
 */
export async function withObservation<T>(
  metric: TrellisDurationMetricName,
  attributes: TrellisDurationMetricAttributes,
  run: () => Promise<T>,
  options?: {
    /** Outcome derivation for a successful result; defaults to `ok`. */
    outcomeOf?: (value: T) => string;
    /** Outcome recorded when the unit throws; defaults to `error`. */
    errorOutcome?: string;
    /** Outcome recorded when an abort signal cancels the unit. */
    signal?: AbortSignal;
  },
): Promise<T> {
  const observation = startObservation(metric, attributes);
  try {
    const value = await run();
    if (options?.signal?.aborted) {
      observation.cancel();
    } else {
      observation.finish(options?.outcomeOf?.(value) ?? "ok");
    }
    return value;
  } catch (error) {
    observation.finish(options?.errorOutcome ?? "error");
    throw error;
  }
}

File: rust/crates/trellis/src/telemetry/lifecycle.rs
//! Exactly-once observation guards for duration and inflight metrics.
//!
//! A guard records one result at its named boundary. Dropping a guard without
//! an explicit finish records the configured cancelled/interrupted outcome, so
//! `?`, panic, or dropped futures cannot silently skip the observation. Guards
//! never change control flow and cannot turn a business failure into a success.

use std::time::{Duration, Instant};

use opentelemetry::KeyValue;

use super::instruments::{add_updown, record_family_duration, DurationFamily, UpDownFamily};

/// Duration observation for one synchronous or asynchronous unit of work.
pub struct Observation {
    started: Instant,
    family: DurationFamily,
    attributes: Vec<KeyValue>,
    drop_outcome: &'static str,
    inflight: Option<InflightGuard>,
    finished: bool,
}

impl Observation {
    /// Starts one un-counted duration observation.
    pub fn start(
        family: DurationFamily,
        attributes: Vec<KeyValue>,
        drop_outcome: &'static str,
    ) -> Self {
        Self {
            started: Instant::now(),
            family,
            attributes,
            drop_outcome,
            inflight: None,
            finished: false,
        }
    }

    /// Starts one duration observation plus an inflight up/down increment.
    pub fn start_counted(
        family: DurationFamily,
        inflight: UpDownFamily,
        attributes: Vec<KeyValue>,
        drop_outcome: &'static str,
    ) -> Self {
        let guard = InflightGuard::acquire(inflight, attributes.clone());
        Self {
            started: Instant::now(),
            family,
            attributes,
            drop_outcome,
            inflight: Some(guard),
            finished: false,
        }
    }

    /// Monotonic elapsed time since the observation started.
    pub fn elapsed(&self) -> Duration {
        self.started.elapsed()
    }

    /// Finishes the observation with one bounded outcome value.
    pub fn finish(self, outcome: &'static str) {
        self.finish_with(outcome, &[]);
    }

    /// Finishes the observation with an outcome and extra attributes.
    pub fn finish_with(mut self, outcome: &'static str, extra: &[KeyValue]) {
        if self.finished {
            return;
        }
        self.finished = true;
        let mut attributes = std::mem::take(&mut self.attributes);
        attributes.push(KeyValue::new("trellis.outcome", outcome));
        attributes.extend_from_slice(extra);
        record_family_duration(self.family, self.started.elapsed(), &attributes);
        if let Some(inflight) = self.inflight.take() {
            drop(inflight);
        }
    }
}

impl Drop for Observation {
    fn drop(&mut self) {
        if self.finished {
            return;
        }
        self.finished = true;
        let mut attributes = std::mem::take(&mut self.attributes);
        attributes.push(KeyValue::new("trellis.outcome", self.drop_outcome));
        record_family_duration(self.family, self.started.elapsed(), &attributes);
        if let Some(inflight) = self.inflight.take() {
            drop(inflight);
        }
    }
}

/// Inflight up/down counter increment released exactly once.
pub struct InflightGuard {
    family: UpDownFamily,
    attributes: Vec<KeyValue>,
    released: bool,
}

impl InflightGuard {
    /// Increments the family and returns the guard that decrements it.
    pub fn acquire(family: UpDownFamily, attributes: Vec<KeyValue>) -> Self {
        add_updown(family, 1, &attributes);
        Self {
            family,
            attributes,
            released: false,
        }
    }

    /// Releases the increment early; dropping has the same effect once.
    pub fn release(mut self) {
        if !self.released {
            self.released = true;
            add_updown(self.family, -1, &self.attributes);
        }
    }
}

impl Drop for InflightGuard {
    fn drop(&mut self) {
        if !self.released {
            self.released = true;
            add_updown(self.family, -1, &self.attributes);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dropped_observations_finish_once() {
        let observation = Observation::start(
            DurationFamily::RpcClient,
            vec![KeyValue::new("trellis.route", "_unknown")],
            "cancelled",
        );
        drop(observation);
    }

    #[test]
    fn explicit_finish_is_idempotent_under_drop() {
        let observation = Observation::start(
            DurationFamily::Cli,
            vec![KeyValue::new("trellis.command", "whoami")],
            "cancelled",
        );
        observation.finish("ok");
    }

    #[test]
    fn inflight_guard_releases_early_once() {
        let guard = InflightGuard::acquire(
            UpDownFamily::RpcServerInflight,
            vec![KeyValue::new("trellis.route", "_unknown")],
        );
        guard.release();
    }
}
